framework/include/userobject/BulkMaterial.h

```diff
@@ -0,0 +1,242 @@
+    if constexpr (I + 1 < sizeof...(Input))
+      copy<I + 1>(nqp);
+  }
+
+public:
+  static InputParameters validParams() { return ElementUserObject::validParams(); }
+
+  template <typename... Names>
+  BulkMaterial(const InputParameters & params, Names &&... names) : ElementUserObject(params)
+  {
+    construct<0, Names...>(std::forward<Names>(names)...);
+  }
+
+  /// override this method to implement the computation on the bulk data
+  virtual void bulkCompute() = 0;
+
+  /// The output type is directly specified as a template parameter
+  typedef Output OutputType;
+  /// The serialized bulk data is stored in a vector
+  typedef std::vector<OutputType> OutputVector;
+
+  /// input data needs to use a flexible tuple type (std::tuple or cuda::std::tuple)
+  typedef typename Tuple::template type<typename Input::type...> InputType;
+  /// The serialized bulk data is stored in a vector
+  typedef std::vector<InputType> InputVector;
+
+  /// get a reference to the output data
+  const OutputVector & getOutputData() const { return _output_data; }
+
+  /// get a reference to the input data
+  const InputVector & getInputData() const { return _input_data; }
+
+  /// check if teh output is fully computed and ready to be fetched
+  bool outputReady() const { return _output_ready; }
+
+  /// get the index for the first qp of a specified element (other qps are at the following indices)
+  std::size_t getIndex(dof_id_type elem_id) const
+  {
+    const auto it = _index_map.find(elem_id);
+    if (it == _index_map.end())
+      mooseError("Element ", elem_id, " was not found in the index map for ", name(), ".");
+    return it->second;
+  }
+
+  /// data structures are initialized here
+  virtual void initialize() final
+  {
+    _index = 0;
+    _output_ready = false;
+  }
+
+  /// All data is automatically gathered here
+  virtual void execute() final
+  {
+    // update index map
+    _index_map[_current_elem->id()] = _index;
+
+    // make sure the input data is sized sufficiently big
+    const auto nqp = _qrule->n_points();
+    if (_input_data.size() < _index + nqp)
+      _input_data.resize(_index + nqp);
+
+    // copy data
+    copy<0>(nqp);
+    _index += nqp;
+  }
+
+  /// Assemble data from all threads
+  virtual void threadJoin(const UserObject & uo) final
+  {
+    // join maps (with index shift)
+    const auto & bm = static_cast<const BulkMaterialType &>(uo);
+    for (const auto & [id, index] : bm._index_map)
+      _index_map[id] = index + _index;
+
+    // make sure we have enough space
+    const auto capacity = _input_data.capacity();
+    if (capacity < _index + bm._index)
+      _input_data.reserve(_index + bm._index);
+
+    // concatenate input data
+    _input_data.insert(std::begin(_input_data) + _index,
+                       std::begin(bm._input_data),
+                       std::begin(bm._input_data) + bm._index);
+    _index += bm._index;
+  }
+
+  /// we call bulkCompute() from here
+  virtual void finalize() final
```

Comment:
I would mark all of these as `override final` (and then you can omit `virtual` if you wish as `override` always implies `virtual`)

Reply:
Yeah, agree. We should probably put that in our styleguide somewhere. I've seen PRs as `virtual` to functions already marked as `override` before.